### ISSUE TITLE
Feature: @Select decorator type safe store selection 

### ISSUE THREAD

**Issue by Silthus** [ngxs/store#160]
I mentioned this before as comment in another issue, but I wanted to properly propose this again.

Would it be possible to overload the `@Select` operator to accept a type safe select function into the provided store.

```
export interface MyStateModel {
    foo: string;
    bar: number;
}

@State<MyStateModel>()
export class MyState {}


@Select(MyState, (state) => state.foo)
foo$: string;
```

**Comment by RobCannon** [comment]
I would like to see that too (or something similar).  The current syntax (where state is of type any) is no better than a string in the selector in terms of refactoring support.

**Comment by deebloo** [comment]
@RobCannon I disagree that it is "no better then a string" but I do agree it could be improved. It would have to be done with a generic. 

```TS
@Select<MyStateModel>(state => state.foo)
foo$: string;
```

which in this case I see as not much better then what you can do now.
```TS
@Select((state: MyStateModel) => state.foo)
foo$: string;
```


what I recommend is to always use the class as the select or a designated selector.

```TS
@Select(MyState)
foo$: string;

@Select(MyState.MySelector)
bar$: string;
```

**Comment by amcdnl** [comment]
Decorators can't imply types in TypeScript right now. There is already an issue tracking that.

**Comment by deebloo** [comment]
@amcdnl I think that issue is different.

**Comment by Silthus** [comment]
@deebloo 

regarding the second case where you mentioned it could be done now, is wrong.

```
@Select((state: MyStateModel) => state.foo) // does not work
foo$: string;
```

The above does not work since the decorator does not select from the given state, but tries to select from the root state. To make it work we must currently use the following syntax loosing type safety:

```
@Select(state => state.mystate.foo)
foo$: string;
```

Wouldn't it be possible to imply the type in a decorator if the state that should be used gets passed as the first argument?

```
@Select(MyState, (state) => state.foo) // as an alternative use MyStateModel?
foo$: string;
```

**Comment by deebloo** [comment]
@Silthus that is correct it would have to be a model of your whole state but that would work.

we can't infer your state model type from the state class. 

I do like that proposed api but I don't think we can make it any more type safe without the use of a generic. If anyone knows a better way I am more then happy to investigate.

```
@Select<SomeModel>(state => state.someProp)
```

**Comment by deebloo** [comment]
@Silthus thoughts?

**Comment by Silthus** [comment]
I think the generics approach does not work with the current version of typescript, also see #75. 
That's why I suggested the workaround with the store or state model class being passed as first argument.

**Comment by deebloo** [comment]
The generics approach will work. It just can't set the type of the property. That is the issue Austin was referring to

**Comment by deebloo** [comment]
The generics approach will work. It just can't set the type of the property. That is the issue Austin was referring to

**Comment by deebloo** [comment]
I think he misunderstood that original issue

**Comment by deebloo** [comment]
The the the generic can't so is select a slice of the state.

**Comment by Silthus** [comment]
Wouldn't it be possible to store the type of the generic state model in the meta data and then get the store by selecting it from the model type in the select generic.

Add the following in the [@State decorator](https://github.com/ngxs/store/blob/master/packages/store/src/state.ts#L22)
```
meta.modelType = typeof ((options.defaults || {}) as T);
```

I don't really understand the code behind the `@Select` decorator so I am not sure if this will work.

**Comment by deebloo** [comment]
@Silthus unless I am way far off I don't think that will work. I would love to see a demo proving me wrong though!

**Comment by Silthus** [comment]
I think you are right.
The best way probably would be like you suggested by using a generic and the store class. This would also have the benefit of not breaking existing code.

**Comment by StevenLiekens** [comment]
So right now there is no way to have `@Select` infer the type of a state model? I'm confused about why this was closed.

**Comment by offwork** [comment]
`@Select(State.getMyData) myData$:Observable<MyData>;` is very useful if you want to constantly check the status changes.
`myData$:Observable<MyData> = store.select<MyData>(State.getMyData).pipe(take(1));` is very useful if you want to check the status change only once.

**Comment by arturovt** [comment]
@offwork 
`selectOnce` is easier to use and read :wink:

**Comment by appleseedexm** [comment]
So since this is closed / ignored I thought I'd post how you can currently solve this according to the solution @deebloo mentioned.

I wished there was a clean implementation at some point but this is still way better than no type-safety.


```javascript
// create a root state model
export interface RootStateModel {
  stateOneName: StateOneModel; // name must be according to state name
  stateTwoName: StateTwoModel;
  ...
}

// define state
export interface StateOneModel {
  ...
}

@State<StateOneModel>({
  name: 'stateOneName', // see RootStateModel interface
  ...
)


// define state two
...
```

Now you can infer types in Selectors like this:

```javascript
// state.stateOne will be type-safe for StateOneModel
@Select((state: RootStateModel) => state.stateOne) terminals$: Observable<StateOneModel>;
```

**Comment by asinitson** [comment]
The trick with `RootStateModel` works! Thanks, @appleseedexm!

Maybe it will help others if I make your example a bit more concrete with single state model:

```typescript
/**
 * Workaround to provide type safety in selectors:
 * https://github.com/ngxs/store/issues/160#issuecomment-521216207
 * each property name has to exactly match respective State name
 * in the decorator.
 */

export class AppStateModel {
  isConnected: boolean;
}

export interface RootStateModel {
  appState: AppStateModel;
}

@State<AppStateModel>({
  name: 'appState',
  defaults: {
    isConnected: false
  }
})
export class AppState {
}
```
And here is how it will look if you use `select` on the `Store` object instead of decorator:

```typescript
  get isConnected() {
    return this.store.select(
      (rootState: RootStateModel) => rootState.appState.isConnected);
  }
```

![image](https://user-images.githubusercontent.com/13285303/75862610-dfc17100-5e07-11ea-8c0f-77eed111a1e3.png)

**Comment by vvolodin** [comment]
decided to try ngxs in a new project, stumbled upon the issue described here. I want type safety everywhere, can't have it in my selectors... bizarre. Found this issue and it's closed without a comment. I guess I'm back to ngrx.

**Comment by markwhitfeld** [comment]
Hi @vvolodin 
Thank you for your comment. I wish you luck with ngrx if that is your choice.

PS. If you would like to look at ngxs again, I would recommend that you look at the https://www.ngxs.io docs where you will discover that ngxs has a number of different ways to write your selectors and to achieve type safety, and that ngxs introduces a paradigm where you do not need a type for the root state. The issue you came across here is very old, and the approach being asked about here (with the inline arrow function for the select) is considered a bad practice. I encourage you to look at the `createSelector` function as well as these utilities I mention here:
https://github.com/ngxs/store/pull/1623#issuecomment-887810699

**Comment by vvolodin** [comment]
Hi @markwhitfeld
Thanks for your reply, I'm afraid the tone of my comment may be too harsh, I was pretty disappointed to go back to ngrx, since I liked the approach used in ngxs.
I will definitely read the docs again, I understand that creating a selector would help with enforcing the types while selecting a slice but correct me if I'm wrong - there's no way for @Select to enforce type checking on the component property it's being used on?

**Comment by markwhitfeld** [comment]
There is, unfortunately, no way to infer types with the `@Select` decorator because of limitations in typescript, but that does not mean that there are no other options. You could inject the store and use `this.store.select(...` or you could use the utilities I made here for a much cleaner and more typesafe approach: https://github.com/ngxs/store/pull/1623#issuecomment-887810699

**Comment by markwhitfeld** [comment]
@Grochni Thank you for your suggestion here.
Unfortunately, it looks like the same approach that we added to NGXS in v3.6.0, which had to be rolled back in v3.7.0 due to the fact that it broke any usages of the `@Select` decorator for private fields. See PR #1623 for the very lengthy discussion and attempts to get around this limitation of typescript. If you think that your approach can work for private fields then I am very interested.

**Comment by Grochni** [comment]
Ah unfortunately I wasn't aware of that specific limitation. As you said, this is of course also the case in my suggested code, as I'm using `keyof`.

**Comment by dmrickey** [comment]
@markwhitfeld Why is this issue closed? Is there a way included in the latest released version to do this now?

**Comment by Y3kRulez** [comment]
I understand that there are means to achieve type safety, but that requires writing additional code (Selectors).
Considering that a state could easily contain 5-10 properties, that's a lot to write.

Has there been any updates to typescript which lets selector to be defined as proposed in the previous comments, without breaking the type safety?

I've been trying NGXS coming from a lightweight state management base service class that we use internally (basically containing a behavior subject to hold the state, a method to partially set a slice of state, and a select method to subscribe to a slice of state). With such approach selecting a slice of state only requires to do: this.stateServ.select(m => m.StateProperty).subscribe(), with full type-safety..i'd really hate to have to declare a selector in the state for every single property (I get that it could be helpful for performance reasons thanks to memoization, but that should not be a requirement..)

**Comment by dmrickey** [comment]
@markwhitfeld
I'm still confused about how this is accomplished with the latest release
